### PR TITLE
#3302 Payment Discount on Orderline, Invoice Candidate 

### PR DESCRIPTION
**Pull request by cristinamghita**
#3302 Payment Discount on Orderline, Invoice Candidate

### CODE REVIEW

File: de.metas.business/src/main/java/org/adempiere/pricing/api/IMDiscountSchemaBL.java [hunk 1]
```diff
@@ -48,7 +52,7 @@
 	 * @param bPartnerFlatDiscount
 	 * @return
 	 */
-	BigDecimal calculateDiscount(I_M_DiscountSchema schema, BigDecimal Qty, BigDecimal Price, int M_Product_ID, int M_Product_Category_ID, BigDecimal bPartnerFlatDiscount);
+	DiscountRequest calculateDiscount(I_M_DiscountSchema schema, BigDecimal Qty, BigDecimal Price, int M_Product_ID, int M_Product_Category_ID, BigDecimal bPartnerFlatDiscount);
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/pricing/api/IMDiscountSchemaBL.java]: as you used it, i this class is a DiscountResult .

More, you could group the method params and create the DiscountRequest.
You will be able to unify all calculateDiscount methods using DiscountRequest.

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceHeaderImpl.java [hunk 2]
```diff
@@ -341,4 +349,61 @@ public BigDecimal calculateTotalNetAmtFromLines()
 
 		return totalNetAmt;
 	}
+
+	@Override
+	public int getC_PaymentTerm_ID()
```
Comment (teosarca) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceHeaderImpl.java]: InvoiceHeaderImpl it's a simple POJO. Please extract this functionality from here, intoduce a field for paymentTermId, set it...

Comment (cristinamghita) [reply de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceHeaderImpl.java]: Ah.. ok
I saw the calculate method.. so.. from there
I will extract.

File: de.metas.business/src/main/java/org/adempiere/pricing/api/DiscountResult.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * #L%
+ */
+
+/**
+ * @author metas-dev <dev@metasfresh.com>
+ *
+ */
+@Builder
+@Getter
+public class DiscountResult
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/pricing/api/DiscountResult.java]: `@Value` pls

File: de.metas.business/src/main/java/org/adempiere/pricing/api/impl/CalculateDiscountCommand.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+/**
+ * @author metas-dev <dev@metasfresh.com>
+ *
+ */
+/* package */ class CalculateDiscountCommand
+{
+
+	final CalculateDiscountRequest request;
+
+	public CalculateDiscountCommand(final CalculateDiscountRequest request)
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/pricing/api/impl/CalculateDiscountCommand.java]: `@NonNull`

File: de.metas.business/src/main/java/org/adempiere/pricing/api/impl/CalculateDiscountCommand.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+	/**
+	 * Check if the instances are empty ( have "" value)
+	 *
+	 * @param instances
+	 * @return true if all the instances are empty, false if at least one is not
+	 */
+	private boolean hasNoValues()
+	{
+		if (request.getInstances() == null || request.getInstances().isEmpty())
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/pricing/api/impl/CalculateDiscountCommand.java]: request.getInstances() is never null

**Pull request merged**